exonum_derive/src/pb_convert.rs

```diff
@@ -0,0 +1,165 @@
+          Ok(Self {
+           #( #our_struct_names: ProtobufConvert::from_pb(pb.#getters().to_owned())?, )*
+          })
+        }
+    }
+}
+
+fn gen_protobuf_convert_to_pb(field_names: &[Ident]) -> impl quote::ToTokens {
+    let setters = field_names
+        .iter()
+        .map(|i| Ident::new(&format!("set_{}", i), Span::call_site()));
+    let our_struct_names = field_names.to_vec();
+
+    quote! {
+        fn to_pb(&self) -> Self::ProtoStruct {
+            let mut msg = Self::ProtoStruct::new();
+            #( msg.#setters(ProtobufConvert::to_pb(&self.#our_struct_names).into()); )*
+            msg
+        }
+    }
+}
+
+fn gen_to_protobuf_impl(
+    name: &Ident,
+    pb_name: &Path,
+    field_names: &[Ident],
+) -> impl quote::ToTokens {
+    let to_pb_fn = gen_protobuf_convert_to_pb(field_names);
+    let from_pb_fn = gen_protobuf_convert_from_pb(field_names);
+
+    quote! {
+        impl ProtobufConvert for #name {
+            type ProtoStruct = #pb_name;
+
+            #to_pb_fn
+            #from_pb_fn
+        }
+    }
+}
+
+fn gen_binary_form_impl(name: &Ident, cr: &quote::ToTokens) -> impl quote::ToTokens {
+    quote! {
+        impl #cr::messages::BinaryForm for #name {
+
+            fn encode(&self) -> std::result::Result<Vec<u8>, _EncodingError> {
+                self.to_pb().write_to_bytes().map_err(|e| _EncodingError::Other(Box::new(e)))
+            }
+
+            fn decode(buffer: &[u8]) -> std::result::Result<Self, _EncodingError> {
+                let mut pb = <Self as ProtobufConvert>::ProtoStruct::new();
+                pb.merge_from_bytes(buffer).map_err(|e| _EncodingError::Other(Box::new(e)))?;
+                Self::from_pb(pb).map_err(|_| "Conversion from protobuf error.".into())
+            }
+        }
+    }
+}
+
+fn gen_storage_traits_impl(name: &Ident, cr: &quote::ToTokens) -> impl quote::ToTokens {
+    quote! {
+        impl #cr::crypto::CryptoHash for #name {
+            fn hash(&self) -> #cr::crypto::Hash {
+                let v = self.to_pb().write_to_bytes().unwrap();
+                #cr::crypto::hash(&v)
+            }
+        }
+
+        impl #cr::storage::StorageValue for #name {
+            fn into_bytes(self) -> Vec<u8> {
+                self.to_pb().write_to_bytes().unwrap()
+            }
+
+            fn from_bytes(value: std::borrow::Cow<[u8]>) -> Self {
+                let mut block = <Self as ProtobufConvert>::ProtoStruct::new();
+                block.merge_from_bytes(value.as_ref()).unwrap();
+                ProtobufConvert::from_pb(block).unwrap()
+            }
+        }
+    }
+}
+
+fn get_field_names(input: &DeriveInput) -> Vec<Ident> {
+    let data = match &input.data {
+        Data::Struct(x) => x,
+        _ => panic!("Protobuf convert can be derived for structs only."),
+    };
+    data.fields
+        .iter()
+        .map(|f| f.ident.clone().unwrap())
+        .collect()
+}
+
+pub fn generate_protobuf_convert(input: TokenStream) -> TokenStream {
```

Comment:
Nit: Inconsistency in method naming - `generate_*` here vs `gen_*` earlier. I'm in favor of a third way, `implement`: `implement_binary_form`, `implement_storage_traits`, etc.